### PR TITLE
Users/yz/fix cors

### PR DESCRIPTION
**Pull request by yangzhi-tech**

### CODE REVIEW

File: server/api/order.go [hunk 1]
```diff
@@ -45,14 +58,119 @@ func CreateOrder(w http.ResponseWriter, r *http.Request) {
...
+		http.Error(w, fmt.Sprintf("More than 1 order found for %s", req.ID), http.StatusInternalServerError)
+		return
+	}
+	order := orders[0]
+	order.Status = req.Status
+	if req.Status == model.StatusSent {
+		order.DepartureTime = int(time.Now().Unix())
+	}
+	if req.Status == model.StatusDeliverd {
+		order.FinishTime = int(time.Now().Unix())
```
Comment (tim-nju) [inline server/api/order.go:163]: Before updating order, we might want to do order status validation here, i.e., comparing previous status with current and judge whether it is a reaonsble status transition. If not, there is bug somewhere (either in the backend or introduced by interaction among front end and backend)>

File: server/payment/client.go [hunk 2]
```diff
@@ -54,6 +50,32 @@ func (c *Client) Pay(price int, desc string, succURL string, cancelURL string) (
 	return resp.ID, resp.InvoiceURL, nil
 }
 
+func (c *Client) GetStatus(url string) (string, error) {
+	httpReq, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return "", err
+	}
+	httpReq.Header.Set("x-api-key", c.apikey)
+	res, err := c.httpClient.Do(httpReq)
```
Comment (tim-nju) [inline server/payment/client.go:59]: Is there convenient lib we can use here for request timeout and retry?

File: server/server.go [hunk 3]
```diff
@@ -16,15 +21,68 @@ import (
...
+	}
+
+}
+
+var AllowedMethods = []string{"GET", "POST", "PUT", "DELETE", "OPTIONS"}
 
-	log.Fatal(srv.ListenAndServe())
+func CORSMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Access-Control-Allow-Origin", "*")
```
Comment (tim-nju) [inline server/server.go]: This looks dangerous. I understand that it is for front end to access (resolving cross-domain issue). Can we limit it to our front end domain only? Thx!

File: server/store/mongo_store.go [hunk 4]
```diff
@@ -43,3 +40,40 @@ func (s *MongoStore) CreateOrder(order *model.Order) error {
 	_, err := collection.InsertOne(context.Background(), order)
 	return err
 }
+
+func (s *MongoStore) ListOrders(status int) ([]*model.Order, error) {
+	collection := s.client.Database(database).Collection(orderCollection)
+	filter := bson.D{{"status", status}}
+	rs, err := collection.Find(context.Background(), filter)
```
Comment (tim-nju) [inline server/store/mongo_store.go]: Same as https://github.com/ec-3/ec-3-shopping/pull/2#discussion_r1536722753

File: server/store/mongo_store.go [hunk 5]
```diff
@@ -43,3 +40,40 @@ func (s *MongoStore) CreateOrder(order *model.Order) error {
...
+	if err != nil {
+		return nil, err
```
Comment (tim-nju) [inline server/store/mongo_store.go:57]: Should we log it before returning?

File: server/api/order.go (hunk 1 above)
Comment (yangzhi-tech) [reply server/api/order.go:163]: Updated.

File: server/payment/client.go (hunk 2 above)
Comment (yangzhi-tech) [reply server/payment/client.go:59]: This is good comment actually.
What about we add some metrics collect latency and succ rate then we know what value to use?
I have concern that adding timeout or retry may make things worse. I am under the impression that the requests take a quite long duration(about 10s).

Comment (yangzhi-tech) [reply server/payment/client.go:59]: I also added a 60 seconds timeout to the client.

File: server/store/mongo_store.go (hunk 5 above)
Comment (yangzhi-tech) [reply server/store/mongo_store.go:57]: Yes. We should. This is on my todo list. I am exploring how to collect log for the web app. Currently it is running in container.

File: server/store/mongo_store.go (hunk 4 above)
Comment (yangzhi-tech) [reply server/store/mongo_store.go]: Updated

**Review by tim-nju — APPROVED** [review]
LGTM. Talked offline and will adress the remaning comments later.

**Pull request merged**